atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTableTest.java

```diff
@@ -159,6 +174,160 @@ public void onceNonNullValueIsReturnedItIsAlwaysReturned() {
+        } else {
+            verify(spiedKvs, never()).checkAndSet(any());
+        }
+        verify(spiedStore, times(102)).put(anyMap());
+    }
+
+    @Test
+    public void coalesceConcurrentGetsForSameTimestampsInPresenceOfFailuresButSucceedsWhenFailuresStop()
+            throws ExecutionException, InterruptedException {
+        setupStagingValues(5);
+
+        ExecutorService writers = Executors.newFixedThreadPool(100);
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        List<Future<ListenableFuture<Long>>> results = new ArrayList<>();
+        int numberOfReads = 10_000;
+        for (int i = 0; i < numberOfReads; i++) {
+            results.add(writers.submit(() -> pueTable.get(random.nextLong(5L))));
+        }
+        results.forEach(future -> assertThatThrownBy(() -> future.get().get())
+                .isInstanceOf(ExecutionException.class)
+                .hasCauseInstanceOf(RuntimeException.class)
+                .hasMessageContaining("Failed to set value"));
+        writers.shutdownNow();
+
+        long checkAndSets = getNumberOfInvocations(spiedKvs, "checkAndSet");
+        long puts = getNumberOfInvocations(spiedStore, "put");
+
+        // assert that some concurrent gets get coalesced but no guarantees due to scheduling
+        if (validating) {
+            assertThat(checkAndSets).isLessThan(numberOfReads / 2);
+            assertThat(checkAndSets).isEqualTo(puts - 1);
+        } else {
+            assertThat(checkAndSets).isEqualTo(0L);
+        }
+
+        spiedStore.stopFailingPuts();
+        assertThat(pueTable.get(0L).get()).isEqualTo(0L);
+        if (validating) {
+            assertThat(getNumberOfInvocations(spiedKvs, "checkAndSet")).isEqualTo(checkAndSets + 1);
+        } else {
+            assertThat(checkAndSets).isEqualTo(0L);
+        }
+        assertThat(getNumberOfInvocations(spiedStore, "put")).isEqualTo(puts + 1);
+    }
+
+    @Test
+    public void noSuperfluousCasOrPuts() {
+        setupStagingValues(50);
+
+        spiedStore.stopFailingPuts();
+        ExecutorService writers = Executors.newFixedThreadPool(100);
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        List<Future<ListenableFuture<Long>>> results = new ArrayList<>();
+        int numberOfReads = 20_000;
+        for (int i = 0; i < numberOfReads; i++) {
+            results.add(writers.submit(() -> pueTable.get(random.nextLong(50L))));
+        }
+        results.forEach(future -> {
+            try {
+                assertThat(future.get().get()).isBetween(0L, 49L);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+        writers.shutdownNow();
+
+        if (validating) {
+            verify(spiedKvs, times(50)).checkAndSet(any());
+        } else {
+            verify(spiedKvs, never()).checkAndSet(any());
+        }
+        verify(spiedStore, times(51)).put(anyMap());
```

Comment:
Might be missing something: this is 51 and not 50, because of the setupStagingValues() call?

Reply:
yes